### ISSUE TITLE
Package leaderelection has data races

### ISSUE THREAD

**Issue by mattmoor** [kubernetes/kubernetes#90952]
**What happened**:

We use the `leaderelection` package downstream in Knative (currently 1.16.x libs), and virtually any unit testing I try with the `LeaderElector` fails under `-race` due to assorted data races.

For example:
```
==================
WARNING: DATA RACE
Write at 0x00c00045d528 by goroutine 23:
  k8s.io/client-go/tools/leaderelection/resourcelock.(*LeaseLock).Get()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/client-go/tools/leaderelection/resourcelock/leaselock.go:41 +0x1b9
  k8s.io/client-go/tools/leaderelection.(*LeaderElector).tryAcquireOrRenew()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:327 +0x18f
  k8s.io/client-go/tools/leaderelection.(*LeaderElector).renew.func1.1.1()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:268 +0x6f

Previous write at 0x00c00045d528 by goroutine 21:
  k8s.io/client-go/tools/leaderelection/resourcelock.(*LeaseLock).Update()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/client-go/tools/leaderelection/resourcelock/leaselock.go:68 +0x504
  k8s.io/client-go/tools/leaderelection.(*LeaderElector).release()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:305 +0x176
  k8s.io/client-go/tools/leaderelection.(*LeaderElector).renew()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:293 +0x1c9
  k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:207 +0x18b

Goroutine 23 (running) created at:
  k8s.io/client-go/tools/leaderelection.(*LeaderElector).renew.func1.1()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:266 +0xa9
  k8s.io/apimachinery/pkg/util/wait.PollImmediateUntil()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:393 +0x38
  k8s.io/client-go/tools/leaderelection.(*LeaderElector).renew.func1()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:264 +0x1ab
  k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x75
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0x108
  k8s.io/apimachinery/pkg/util/wait.Until()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x182
  k8s.io/client-go/tools/leaderelection.(*LeaderElector).renew()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:261 +0xcc
  k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run()
      /Users/mattmoor/go/src/knative.dev/pkg/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:207 +0x18b

Goroutine 21 (finished) created at:
  knative.dev/pkg/leaderelection.TestRace()
      /Users/mattmoor/go/src/knative.dev/pkg/leaderelection/race_test.go:110 +0xb50
  testing.tRunner()
      /usr/local/Cellar/go/1.14.2_1/libexec/src/testing/testing.go:991 +0x1eb
==================
```

**What you expected to happen**:

I'd expect folks to be able to safely use the `leaderelection` package under `-race`.

**How to reproduce it (as minimally and precisely as possible)**:

I refactored my test to one that just depends on K8s libs here, and it is failing with `-race`:

```go
package leaderelection

import (
	"context"
	"testing"
	"time"

	"k8s.io/apimachinery/pkg/runtime"
	fakekube "k8s.io/client-go/kubernetes/fake"
	ktesting "k8s.io/client-go/testing"
	"k8s.io/client-go/tools/leaderelection"
	"k8s.io/client-go/tools/leaderelection/resourcelock"
)

func TestRace(t *testing.T) {
	kc := fakekube.NewSimpleClientset()

	// Detect promotion
	promoted := make(chan struct{})
	onStart := func(context.Context) { close(promoted) }

	// Detect demotion
	demoted := make(chan struct{})
	onStop := func() { close(demoted) }

	// Detect lease creation
	created := make(chan struct{})
	kc.PrependReactor("create", "leases",
		func(action ktesting.Action) (bool, runtime.Object, error) {
			close(created)
			return false, nil, nil
		},
	)

	// Detect lease renewal
	updated := make(chan struct{})
	kc.PrependReactor("update", "leases",
		func(action ktesting.Action) (bool, runtime.Object, error) {
			// Only close update once.
			select {
			case <-updated:
			default:
				close(updated)
			}
			return false, nil, nil
		},
	)

	// Create the resource lock.
	rl, err := resourcelock.New("leases",
		"repro-system",
		"name-of-lock",
		kc.CoreV1(),
		kc.CoordinationV1(),
		resourcelock.ResourceLockConfig{
			Identity: "unique-identity-ftw",
		})
	if err != nil {
		t.Fatalf("resourcelock.New() = %v", err)
	}

	// Create the elector around the above resource lock.
	le, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
		Lock:          rl,
		LeaseDuration: 15 * time.Second,
		RenewDeadline: 10 * time.Second,
		RetryPeriod:   2 * time.Second,
		Callbacks: leaderelection.LeaderCallbacks{
			OnStartedLeading: onStart,
			OnStoppedLeading: onStop,
		},
		ReleaseOnCancel: true,
		Name:            "component",
	})
	if err != nil {
		t.Fatalf("leaderelection.NewLeaderElector() = %v", err)
	}

	// We shouldn't see leases until we Run the elector.
	select {
	case <-promoted:
		t.Error("Got promoted, want no actions.")
	case <-demoted:
		t.Error("Got demoted, want no actions.")
	case <-created:
		t.Error("Got created, want no actions.")
	case <-updated:
		t.Error("Got updated, want no actions.")
	default:
	}

	// Run the elector.
	ctx, cancel := context.WithCancel(context.Background())
	go le.Run(ctx)

	// We should see the lease created shortly thereafter.
	select {
	case <-created:
		// We expect the lease to be created.
	case <-time.After(1 * time.Second):
		t.Fatal("Timed out waiting for lease creation.")
	}
	// We should be promoted as well.
	select {
	case <-promoted:
		// We expect to have been promoted.
	case <-time.After(1 * time.Second):
		t.Fatal("Timed out waiting for promotion.")
	}

	// Cancelling the context should cause us to give up leadership.
	cancel()

	// We should see the lease updated (at least for release, but possibly due to renewal)
	select {
	case <-updated:
		// We expect the lease to be updated.
	case <-time.After(1 * time.Second):
		t.Fatal("Timed out waiting for lease update.")
	}
	// On cancellation, we should be demoted.
	select {
	case <-demoted:
		// We expect to have been demoted.
	case <-time.After(1 * time.Second):
		t.Fatal("Timed out waiting for demotion.")
	}
}

```

**Comment by tedyu** [comment]
[le-race-test.txt](https://github.com/kubernetes/kubernetes/files/4605983/le-race-test.txt)
The above is based on the test you gave.

I ran the test in Goland several times but didn't see the race.

Also the line numbers in the trace don't match what master branch has.

Mind running on master branch and updating the trace if you reproduce it ?

thanks

**Comment by tedyu** [comment]
The write race seems to be between Get():
```
	record := LeaseSpecToLeaderElectionRecord(&ll.lease.Spec)
```
and Update():
```
	ll.lease.Spec = LeaderElectionRecordToLeaseSpec(&ler)
```

**Comment by mattmoor** [comment]
On my phone now, but there were multiple flavors of race uncovered.  There was also a write to write race between release and the renew code.

**Comment by dprotaso** [comment]
The PR https://github.com/kubernetes/kubernetes/pull/87899 removed the extra go routines causing the race - so this is fixed in the 1.18 client libs

**Comment by dims** [comment]
excellent thanks for tracking this down @dprotaso 

/close
